flow/view_slicer.cc

```diff
@@ -0,0 +1,116 @@
+  std::unordered_map<int64_t, SkRect> overlay_layers;
+
+  auto current_frame_view_count = composition_order.size();
+
+  // Restore the clip context after exiting this method since it's changed
+  // below.
+  DlAutoCanvasRestore save(background_canvas, /*do_save=*/true);
+
+  for (size_t i = 0; i < current_frame_view_count; i++) {
+    int64_t view_id = composition_order[i];
+    EmbedderViewSlice* slice = slices.at(view_id).get();
+    if (slice->canvas() == nullptr) {
+      continue;
+    }
+
+    slice->end_recording();
+
+    SkRect full_joined_rect = SkRect::MakeEmpty();
+
+    // Determinate if Flutter UI intersects with any of the previous
+    // platform views stacked by z position.
+    //
+    // This is done by querying the r-tree that holds the records for the
+    // picture recorder corresponding to the flow layers added after a platform
+    // view layer.
+    for (int j = i; j >= 0; j--) {
+      int64_t current_view_id = composition_order[j];
+      auto maybe_rect = view_rects.find(current_view_id);
+      FML_DCHECK(maybe_rect != view_rects.end());
+      if (maybe_rect == view_rects.end()) {
+        continue;
+      }
+
+      SkRect current_view_rect = maybe_rect->second;
+      const SkIRect rounded_in_platform_view_rect = current_view_rect.roundIn();
+
+      // Each rect corresponds to a native view that renders Flutter UI.
+      std::vector<SkIRect> intersection_rects =
+          slice->region(current_view_rect).getRects();
+
+      // Ignore intersections of single width/height on the edge of the platform
+      // view.
+      // This is to address the following performance issue when interleaving
+      // adjacent platform views and layers: Since we `roundOut` both platform
+      // view rects and the layer rects, as long as the coordinate is
+      // fractional, there will be an intersection of a single pixel width (or
+      // height) after rounding out, even if they do not intersect before
+      // rounding out. We have to round out both platform view rect and the
+      // layer rect. Rounding in platform view rect will result in missing pixel
+      // on the intersection edge. Rounding in layer rect will result in missing
+      // pixel on the edge of the layer on top of the platform view.
+      for (auto it = intersection_rects.begin(); it != intersection_rects.end();
+           /*no-op*/) {
+        // If intersection_rect does not intersect with the *rounded in*
+        // platform view rect, then the intersection must be a single pixel
+        // width (or height) on edge.
+        if (!SkIRect::Intersects(*it, rounded_in_platform_view_rect)) {
+          it = intersection_rects.erase(it);
+        } else {
+          ++it;
+        }
+      }
+
+      // Limit the number of native views, so it doesn't grow forever.
+      //
+      // In this case, the rects are merged into a single one that is the union
+      // of all the rects.
```

Comment:
iOS used to have 2 overlays per platform view as the limit. Do you have any context why we picked 2?

Reply:
as far as I can tell it was aribtrarily chosen

Reply:
Wondering if there are any benefit of having multiple overlays despite of the performance overhead. Otherwise it's odd that we picked 2 before. @cbracken do you recall?

Reply:
its possible that at one point the overlay surfaces weren't screen sized, so shrinking the overlays as small as possible improved performance. overlay sizes aren;t stable however, which is what lead to using the frame size in all cases - which defeats the purpose of minimizing overlay size.

Reply:
Gotcha. Thanks! Is this limit the main difference (only difference?) between our ios and android implementations?

Reply:
That, and the android implementation had already fixed the subpixel overlap issue 🫠

Reply:
Yup

Reply:
Gotcha. This PR should move the metrics on both iOS (by reducing the limit from 2 to 1) and Android (by fixing the 1 px overlay). Do you think I should make our ads benchmark work on Android (should be a quick change? CC @jmagman correct me if it's not), so that we can measure both?

Reply:
Almost all Android PVs will use TLHC by default, which does not use platform view slicing. There is work this year to make android PVs work more like iOS by default, but without any of the terrible downsides of the current approach.

Reply:
This means we don't need the Android benchmark to run for this PR (since it's not the implementation used in Android yet), but we may need it later down the road, right? (https://github.com/flutter/flutter/issues/152200)